src/opf/sdpwrm.jl

```diff
@@ -0,0 +1,300 @@
+    @expression(model, wf[e in 1:E], WR[bus_fr[e], bus_fr[e]])
+    @expression(model, wt[e in 1:E], WR[bus_to[e], bus_to[e]])
+    @expression(model, wr[e in 1:E], WR[bus_fr[e], bus_to[e]])
+    @expression(model, wi[e in 1:E], WI[bus_fr[e], bus_to[e]])
+
+    # Ohm's law
+    @constraint(model, ohm_pf[e in 1:E],
+        branch_status[e] * ( gff[e] * wf[e] + gft[e] * wr[e] + bft[e] * wi[e]) - pf[e] == 0
+    )
+    @constraint(model, ohm_qf[e in 1:E],
+        branch_status[e] * (-bff[e] * wf[e] - bft[e] * wr[e] + gft[e] * wi[e]) - qf[e] == 0
+    )
+    @constraint(model, ohm_pt[e in 1:E],
+        branch_status[e] * ( gtt[e] * wt[e] + gtf[e] * wr[e] - btf[e] * wi[e]) - pt[e] == 0
+    )
+    @constraint(model, ohm_qt[e in 1:E],
+        branch_status[e] * (-btt[e] * wt[e] - btf[e] * wr[e] - gtf[e] * wi[e]) - qt[e] == 0
+    )
+
+    # Thermal limit
+    @constraint(model, sm_fr[e in 1:E], [smax[e], pf[e], qf[e]] in SecondOrderCone())
+    @constraint(model, sm_to[e in 1:E], [smax[e], pt[e], qt[e]] in SecondOrderCone())
+
+    # Voltage angle difference limit
+    @constraint(model, va_diff_lb[e in 1:E], branch_status[e] * wi[e] - branch_status[e] * tan(dvamin[e]) * wr[e] >= 0)
+    @constraint(model, va_diff_ub[e in 1:E], branch_status[e] * wi[e] - branch_status[e] * tan(dvamax[e]) * wr[e] <= 0)
+    
+    # PSD constraint
+    @constraint(model, S, [WR WI; -WI WR] in PSDCone())
+    
+    #
+    #   III. Objective
+    #
+    l, u = extrema(c2)
+    (l == u == 0.0) || @warn "Data $(data.case) has quadratic cost terms; those terms are being ignored"
+    @objective(model,
+        Min,
+        sum(c1[g] * pg[g] + c0[g] for g in 1:G if gen_status[g])
+    )
+
+    return OPFModel{SDPOPF}(data, model)
+end
+
+function extract_primal(opf::OPFModel{SDPOPF})
+    model = opf.model
+    T = JuMP.value_type(typeof(model))
+
+    data = opf.data
+
+    N, E, G = data.N, data.E, data.G
+
+    primal_solution = Dict{String,Any}(
+        # bus
+        "w" => zeros(T, N),
+        # generator
+        "pg" => zeros(T, G),
+        "qg" => zeros(T, G),
+        # branch
+        "wr" => zeros(T, E),
+        "wi" => zeros(T, E),
+        "pf" => zeros(T, E),
+        "qf" => zeros(T, E),
+        "pt" => zeros(T, E),
+        "qt" => zeros(T, E),
+    )
+    if has_values(model)
+        # bus
+        primal_solution["w"] = value.([model[:WR][i, i] for i in 1:N])
+
+        # generator
+        primal_solution["pg"] = value.(model[:pg])
+        primal_solution["qg"] = value.(model[:qg])
+
+        # branch
+        # By construction of the PSD matrix, WR and WI are defined for buspairs instead of branches.
+        # For convenience, they are stored for branches.
+        # The same values will be repeated if there are multiple branches between a bus pair.
+        # Only extract WR and WI entries that correspond to branches (necessary for power flow computation)
+        # Other entries are not extracted as they are dense
```

Comment:
I would like to have some more background for this, in the documentation plz

Reply:
From in-person discussion:
* the `W` matrix is dense, so we are indeed only exporting a partial primal solution
* diagonal entries are exported in the `w` variable
* off-diagonal entries are exported for every pair of buses that are connected by at least a branch
* the other off-diagonal entries do not appear anywhere in the problem, except in the definition of the PSD constraint

Therefore, while we do not export a complete PSD solution, one can be recovered by solving a PSD completion problem. The completion problem consists in finding a `W` matrix whose diagonal entries are equal to `w` and off-diagonal entries corresponding to bus pairs are equal to `wr`/`wi` as extracted. This can be done by solving a feasibility PSD problem, or by a dedicated completion algorithm.